p2p/simulations/mocker.go

```diff
@@ -0,0 +1,194 @@
+
+//initialize the list of mockers
+func initializeMockers() {
+	//create the map
+	mockerList = make(map[string]func(net *Network, quit chan struct{}, nodeCount int))
+	//build the map
+	mockerList["startStop"] = startStop
+	mockerList["probabilistic"] = probabilistic
+	mockerList["boot"] = boot
+
+	mockersInitialized = true
+}
+
+//Lookup a mocker by its name, returns the mockerFn
+func LookupMocker(mockerType string) func(net *Network, quit chan struct{}, nodeCount int) {
+	if !mockersInitialized {
+		initializeMockers()
+	}
+	return mockerList[mockerType]
+}
+
+//Get a list of mockers (keys of the map)
+//Useful for frontend to build available mocker selection
+func GetMockerList() []string {
+	if !mockersInitialized {
+		initializeMockers()
+	}
+	list := make([]string, 0, len(mockerList))
+	for k, _ := range mockerList {
+		list = append(list, k)
+	}
+	return list
+}
+
+//The boot mockerFn only connects the node in a ring and doesn't do anything else
+func boot(net *Network, quit chan struct{}, nodeCount int) {
+	connectNodesInRing(net, nodeCount)
+}
+
+//The startStop mockerFn stops and starts nodes in a defined period (ticker)
+func startStop(net *Network, quit chan struct{}, nodeCount int) {
+	nodes := connectNodesInRing(net, nodeCount)
+	tick := time.NewTicker(10 * time.Second)
+	defer tick.Stop()
+	for {
+		select {
+		case <-quit:
+			log.Info("Terminating simulation loop")
+			return
+		case <-tick.C:
+			id := nodes[rand.Intn(len(nodes))]
+			go func() {
+				log.Info("stopping node", "id", id)
+				if err := net.Stop(id); err != nil {
+					log.Error("error stopping node", "id", id, "err", err)
+					return
+				}
+
+				time.Sleep(3 * time.Second)
+
+				log.Debug("starting node", "id", id)
+				if err := net.Start(id); err != nil {
+					log.Error("error starting node", "id", id, "err", err)
+					return
+				}
+			}()
+		}
+	}
+}
+
+//The probabilistic mocker func has a more probabilistic pattern
+//(the implementation could probably be improved):
+//nodes are connected in a ring, then a varying number of random nodes is selected,
+//mocker then stops and starts them in random intervals, and continues the loop
+func probabilistic(net *Network, quit chan struct{}, nodeCount int) {
+	nodes := connectNodesInRing(net, nodeCount)
+	for {
+		select {
+		case <-quit:
+			log.Info("Terminating simulation loop")
+			return
+		default:
+		}
+		var lowid, highid int
+		var wg sync.WaitGroup
+		randWait := rand.Intn(5000) + 1000
+		rand1 := rand.Intn(9)
+		rand2 := rand.Intn(9)
+		if rand1 < rand2 {
+			lowid = rand1
+			highid = rand2
+		} else if rand1 > rand2 {
+			highid = rand1
+			lowid = rand2
+		} else {
+			if rand1 == 0 {
+				rand2 = 9
+			} else if rand1 == 9 {
+				rand1 = 0
+			}
+			lowid = rand1
+			highid = rand2
+		}
+		var steps = highid - lowid
+		wg.Add(steps)
+		for i := lowid; i < highid; i++ {
+			select {
+			case <-quit:
+				log.Info("Terminating simulation loop")
+				return
+			default:
```

Comment:
The sleep should be done in this select.

Reply:
yes with a `time.After` chan